crates/ruff_python_parser/src/parser/expression.rs

```diff
@@ -1331,6 +1331,62 @@ impl<'src> Parser<'src> {
 
         self.expect(TokenKind::FStringEnd);
 
+        // test_ok pep701_f_string_py312
+        // # parse_options: {"target-version": "3.12"}
+        // f'Magic wand: { bag['wand'] }'     # nested quotes
+        // f"{'\n'.join(a)}"                  # escape sequence
+        // f'''A complex trick: {
+        //     bag['bag']                     # comment
+        // }'''
+        // f"{f"{f"{f"{f"{f"{1+1}"}"}"}"}"}"  # arbitrary nesting
+
+        // test_err pep701_f_string_py311
+        // # parse_options: {"target-version": "3.11"}
+        // f'Magic wand: { bag['wand'] }'     # nested quotes
+        // f"{'\n'.join(a)}"                  # escape sequence
+        // f'''A complex trick: {
+        //     bag['bag']                     # comment
+        // }'''
+        // f"{f"{f"{f"{f"{f"{1+1}"}"}"}"}"}"  # arbitrary nesting
+
+        // the inner variant here doesn't matter, just checking if PEP 701 f-strings are supported
+        if UnsupportedSyntaxErrorKind::Pep701FString(FStringKind::NestedQuote)
+            .is_unsupported(self.options.target_version)
+        {
+            let quote_str = flags.quote_str();
+            for expr in elements.expressions() {
+                if let Some(slash_index) = self.source[expr.range].find('\\') {
+                    let Ok(slash_index) = TextSize::try_from(slash_index) else {
+                        continue;
+                    };
+                    self.add_unsupported_syntax_error(
+                        UnsupportedSyntaxErrorKind::Pep701FString(FStringKind::Backslash),
+                        TextRange::at(expr.range.start() + slash_index, TextSize::from(1)),
+                    );
+                };
+
+                if let Some(comment_index) = self.source[expr.range].find('#') {
```

Comment:
I think this will give you false positive when a nested valid string inside an f-string contains a `#` character:
```py
❯ uv run --python=3.9 --no-project python -q                
>>> f"outer {'# not a comment'}"
'outer # not a comment'
```

Reply:
And, in the format specifier:
```py
f'outer {x:{"# not a comment"} }'
```